src/overmap.cpp

```diff
@@ -938,6 +963,81 @@ void apply_region_overlay(JsonObject &jo, regional_settings &region)
 
 }
 
+const overmap_special_terrain &overmap_special::get_terrain_at( const tripoint &p ) const
+{
+    for( const auto &elem : terrains ) {
+        if( elem.p == p ) {
+            return elem;
+        }
+    }
+    static const overmap_special_terrain null_terrain;
+    return null_terrain;
+}
+
+bool overmap_special::requires_existing_road() const
+{
+    return locations.count( "by_road" ) > 0;
+}
+
+void overmap_special::finalize()
+{
+    // Check locations for validity.
+    for( const auto &elem : locations ) {
+        if( overmap_locations.count( elem ) == 0 ) {
+            debugmsg( "Overmap special \"%s\" has invalid location \"%s\".",
+                      id.c_str(), elem.c_str() );
+        }
+    }
+    // Update and check terrains and connections.
+    connections.clear();
+    std::set<tripoint> points;
+
+    for( const auto &elem : terrains ) {
+        if( !elem.terrain.is_valid() ) {
+            debugmsg( "Invalid terrain \"%s\" in overmap special \"%s\".",
+                      elem.terrain.c_str(), id.c_str() );
+            continue;
+        }
+
+        if( points.count( elem.p ) > 0 ) {
+            debugmsg( "In overmap special \"%s\", point [%d,%d,%d] is duplicated.",
+                      id.c_str(), elem.p.x, elem.p.y, elem.p.z );
+        } else {
+            points.insert( elem.p );
+        }
+
+        if( !elem.connect ) {
+            continue;
+        }
+
+        bool found = false;
+
+        for( const auto &n : { NORTH, SOUTH, EAST, WEST } ) {
```

Comment:
I had inlined it [at some point](https://github.com/CleverRaven/Cataclysm-DDA/pull/18982/commits/1379c92210ee4184a140f12c25fcca2ba1c8cbac#diff-90df0fb487770203c1f3f21de7b1b6bdL966), but didn't see any ill effects. I wonder if there's a modern compiler that would stumble here.